File: app/src/main/java/luigi/wal/walgif/GIFWallpaperService.java
package luigi.wal.walgif;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Movie;
import android.graphics.Rect;
import android.os.Handler;
import android.service.wallpaper.WallpaperService;
import android.util.Log;
import android.view.SurfaceHolder;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Random;
import java.util.concurrent.LinkedBlockingDeque;

import io.socket.client.IO;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;


public class GIFWallpaperService extends WallpaperService {


    @Override
    public WallpaperService.Engine onCreateEngine() {
        try {
            String[] list = getResources().getAssets().list("gif");
            Log.d("GIF", "pickProfileImage: ");
            Random random = new Random();
            int rand = random.nextInt(list.length);
            InputStream open = getResources().getAssets().open("gif/" + list[rand]);
            Movie movie = Movie.decodeStream(open);

            List<Movie> movies = new ArrayList<>();
            movies.add(movie);
            return new GIFWallpaperEngine(movies);
        } catch (IOException e) {
            Log.d("GIF", "Could not load asset");
            return null;
        }


    }

    private class GIFWallpaperEngine extends WallpaperService.Engine {
        private static final int CAPACITY = 8;//WILL BECOME A CUSTOM PROPERTY
        private final int frameDuration = 100;//ARBITRARY NUMBER
        private final LinkedBlockingDeque<Movie> movies;
        private final Emitter.Listener listener = new Emitter.Listener() {

            @Override
            public void call(Object... args) {
                JSONObject obj = (JSONObject) args[0];
                try {
                    String gifUrl = (String) obj.get("gif");
                    Movie movieFromURL = getMovieFromURL(gifUrl);//TODO move this into a different thread
                    if (movieFromURL != null) {
                        movies.poll();
                        movies.offer(movieFromURL);

                        Log.d("GIF", "call() called with: args = [" + Arrays.toString(args) + "]" + new Date());
                    }
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            }

        };
        private SurfaceHolder holder;

        private boolean visible;
        private Handler handler;
        /**
         * FIELDS BELLOW
         * THIS WILL BECOME FUNCTIONS
         * OR WILL BE DYNAMICALLY COMPUTED
         */
        private int quarterYMax;
        private int halfXMax;
        private int yMax;
        private int xMax;
        private int cycle;
        private int seconds;
        private Runnable drawGIF = this::draw;

        private GIFWallpaperEngine(List<Movie> pMovies) {
            this.movies = new LinkedBlockingDeque<>(CAPACITY);
            movies.addAll(pMovies);
            handler = new Handler();
            try {
                Socket socket = IO.socket("http://archillect.com/");
                socket.on("gif", listener);
                socket.connect();
            } catch (URISyntaxException e) {
                e.printStackTrace();
            }
        }


        private void draw() {

            if (visible) {

                debug();

                Canvas canvas = holder.lockCanvas();
                canvas.save();
                xMax = canvas.getWidth();
                yMax = canvas.getHeight();
                halfXMax = xMax / 2;
                quarterYMax = yMax / 4;
                //bad code... sorry

                //first picture
                int firstX = 0;
                int firstY = 0;
                //second picture
                int secondX = 0;
                int secondY = quarterYMax;
                //third picture
                int thirdX = 0;
                int thirdY = 2 * quarterYMax;
                //fourth picture
                int fourthX = 0;
                int fourthY = 3 * quarterYMax;
                //fifth picture
                int fifthX = halfXMax;
                int fifthY = 0;
                //sixth picture
                int sixthX = halfXMax;
                int sixthY = quarterYMax;
                //seventh picture
                int seventhX = halfXMax;
                int seventhY = 2 * quarterYMax;
                //eighth picture
                int eighthX = halfXMax;
                int eighthY = 3 * quarterYMax;


                Movie movie = movies.peek();
                if (movie != null) {
                    dostuff(canvas, movie, firstX, firstY);
                    //    dostuff(canvas, movie, secondX, secondY);
                    dostuff(canvas, movie, thirdX, thirdY);
                  /*  dostuff(canvas, movie, fourthX, fourthY);
                    dostuff(canvas, movie, fifthX, fifthY);
                    dostuff(canvas, movie, sixthX, sixthY);
                    dostuff(canvas, movie, seventhX, seventhY);
                    dostuff(canvas, movie, eighthX, eighthY);*/
                    if (movie.duration() > 0) {
                        movie.setTime((int) (System.currentTimeMillis() % movie.duration()));
                    }
                } else {
                    Log.i("GIF", "draw: ko");
                    //insert base gif
                }
                canvas.restore();

                holder.unlockCanvasAndPost(canvas);

                handler.removeCallbacks(drawGIF);
                handler.postDelayed(drawGIF, frameDuration);


            }
        }

        private void debug() {
            int prevSecond = new Date().getSeconds();
            if (seconds != prevSecond) {
                Log.i("GIF", "draw:  cycle = " + cycle + " seconds = " + this.seconds);
                this.seconds = prevSecond;
                cycle = 0;
            }
            cycle += 1;
        }

        private void dostuff(Canvas canvas, Movie movie, int left, int top) {
            //TODO width and height must be > 0
            Bitmap movieBitmap = Bitmap.createBitmap(movie.width(), movie.height(),
                    Bitmap.Config.ARGB_8888);
            Canvas movieCanvas = new Canvas(movieBitmap);
            movie.draw(movieCanvas, 0, 0);
            Rect src = new Rect(0, 0, movie.width(), movie.height());
            Rect dst = new Rect(left, top, left + 2 * halfXMax, top + 2 * quarterYMax);
            canvas.drawBitmap(movieBitmap, src, dst, null);
        }

        @Override
        public void onVisibilityChanged(boolean visible) {
            this.visible = visible;
            if (visible) {
                handler.post(drawGIF);
            } else {
                handler.removeCallbacks(drawGIF);
            }
        }


        @Override
        public void onCreate(SurfaceHolder surfaceHolder) {
            super.onCreate(surfaceHolder);
            this.holder = surfaceHolder;
        }

        @Override
        public void onDestroy() {
            super.onDestroy();
            handler.removeCallbacks(drawGIF);
        }


        private Movie getMovieFromURL(String src) {
            try {
                java.net.URL url = new java.net.URL(src);
                HttpURLConnection connection = (HttpURLConnection) url
                        .openConnection();
                connection.setDoInput(true);
                connection.connect();
                InputStream input = connection.getInputStream();
                return Movie.decodeStream(input);
            } catch (IOException e) {
                e.printStackTrace();
                return null;
            }
        }


    }
}

File: app/src/main/java/luigi/wal/walgif/GIFWallpaperService.kt
package luigi.wal.walgif

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Movie
import android.graphics.Rect
import android.os.Handler
import android.service.wallpaper.WallpaperService
import android.util.Log
import android.view.SurfaceHolder
import io.socket.client.IO
import io.socket.client.Socket
import io.socket.emitter.Emitter
import luigi.wal.walgif.utils.RectangleCalculator
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URISyntaxException
import java.util.*
import java.util.concurrent.*


class GIFWallpaperService : WallpaperService() {
    private var gifAssets: Array<String>? = null

    private val mRandom = Random()

    override fun onCreateEngine(): WallpaperService.Engine? {
        return try {
            gifAssets = resources.assets.list("gif")
            val movies = pickRandomDefaultMovies(4)
            GIFWallpaperEngine(movies)
        } catch (e: IOException) {
            Log.d(GIF_TAG, "Could not load asset")
            null
        }
    }

    @Throws(IOException::class)
    private fun pickRandomDefaultMovies(numberOfMovie: Int): List<Movie> {
        val movies = ArrayList<Movie>()
        for (i in 0 until numberOfMovie) {
            Log.d(GIF_TAG, "pickProfileImage: ")
            movies.add(pickRandomMovie())
        }
        return movies
    }

    @Throws(IOException::class)
    private fun pickRandomMovie(): Movie {
        val rand = mRandom.nextInt(gifAssets?.size!!)
        val open = resources.assets.open("gif/" + gifAssets!![rand])
        return Movie.decodeStream(open)
    }

    private inner class GIFWallpaperEngine internal constructor(pMovies: List<Movie>) : WallpaperService.Engine() {
        private val frameDuration = 100//Practical NUMBER
        private var movies: LinkedBlockingDeque<Movie> = LinkedBlockingDeque(CAPACITY)
        private val listener = Emitter.Listener { args ->
            val obj = args[0] as JSONObject
            try {
                val gifUrl = obj.get("gif") as String
                val movieFromURL = getMovieFromURL(gifUrl)
                if (movieFromURL != null) {
                    movies.poll()
                    movies.offer(movieFromURL)

                    Log.d(GIF_TAG, "call() called with: args = [" + Arrays.toString(args) + "]" + Date())
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
        private var holder: SurfaceHolder? = null
        private var visible: Boolean = false
        private val handler: Handler
        /**
         * FIELDS BELLOW
         * THIS WILL BECOME FUNCTIONS
         * OR WILL BE DYNAMICALLY COMPUTED
         */


        private var cycle: Int = 0
        private var seconds: Int = 0
        private var rectWidth: Int = 0
        private var rectHeigth: Int = 0
        private var mRectanglePoints: List<Pair<Int, Int>>? = null
        private var settingsHasChanged: Boolean = false
        private val drawGIF = Runnable { this.draw() }
        private var mSocket: Socket? = null
        private val mScheduler: ScheduledExecutorService? = Executors.newSingleThreadScheduledExecutor()
        private var mScheduledFutureStartOn: ScheduledFuture<*>? = null
        private var mScheduledFutureStartOff: ScheduledFuture<*>? = null

        init {
            settingsHasChanged = true
            movies.addAll(pMovies)
            handler = Handler()
            try {
                mSocket = IO.socket("http://archillect.com/")
                startListeningArchillect()
            } catch (e: URISyntaxException) {
                Log.e(GIF_TAG, "GIFWallpaperEngine: ", e)
            }

        }

        private fun startListeningArchillect() {

            val startOn = Runnable {
                mSocket?.on(SOCKET_EVENT_NAME, listener)
                Log.i(GIF_TAG, "  mSocket.on(\"gif\", listener);: on")
            }

            mScheduledFutureStartOn = mScheduler?.scheduleWithFixedDelay(startOn, 0, 15, TimeUnit.SECONDS)
            val startOff = Runnable {
                mSocket?.off(SOCKET_EVENT_NAME)
                Log.i(GIF_TAG, "mSocket.off(\"gif\");")
            }
            mScheduledFutureStartOff = mScheduler?.scheduleWithFixedDelay(startOff, 5, 15, TimeUnit.SECONDS)
            mSocket?.connect()
        }


        private fun draw() {

            if (visible) {
                // debug();
                val canvas = holder?.lockCanvas()
                if (canvas != null) {
                    canvas.save()


                    checkSettingsHasChanged(canvas)
                    val moviesClone = LinkedList(movies)
                    mRectanglePoints?.forEach { (first, second) -> newMovie(moviesClone, canvas, first, second) }
                    moviesClone.clear()

                    canvas.restore()

                    holder?.unlockCanvasAndPost(canvas)
                }
                handler.removeCallbacks(drawGIF)
                handler.postDelayed(drawGIF, frameDuration.toLong())


            }
        }

        private fun checkSettingsHasChanged(pCanvas: Canvas) {
            if (settingsHasChanged) {
                val xMax = pCanvas.width
                val yMax = pCanvas.height
                mRectanglePoints = RectangleCalculator.getRectanglePoints(xMax, yMax, DIMENSION)
                val (first, second) = RectangleCalculator.getRectDimension(xMax, yMax, DIMENSION)
                rectWidth = first
                rectHeigth = second
                settingsHasChanged = false
            }
        }

        private fun newMovie(pMoviesClone: LinkedList<Movie>, pCanvas: Canvas, x: Int, y: Int) {
            try {
                var movie: Movie? = pMoviesClone.poll()
                if (movie == null) {
                    Log.i(GIF_TAG, "draw: no movie available")
                    movie = pickRandomMovie()
                }
                drawMovie(pCanvas, x, y, movie)
            } catch (pE: IOException) {
                Log.e(GIF_TAG, "newMovie: ", pE)
            }

        }


        private fun drawMovie(pCanvas: Canvas, pPosX: Int, pPosY: Int, pMovie: Movie) {
            drawBitmapMovie(pCanvas, pMovie, pPosX, pPosY)

            if (pMovie.duration() > 0) {
                pMovie.setTime((System.currentTimeMillis() % pMovie.duration()).toInt())
            }
        }

        private fun drawBitmapMovie(canvas: Canvas, movie: Movie, left: Int, top: Int) {

            val movieBitmap = Bitmap.createBitmap(movie.width(), movie.height(),
                    Bitmap.Config.ARGB_8888)
            val movieCanvas = Canvas(movieBitmap)
            movie.draw(movieCanvas, 0f, 0f)
            val src = Rect(0, 0, movie.width(), movie.height())
            val dst = Rect(left, top, left + rectWidth, top + rectHeigth)
            canvas.drawBitmap(movieBitmap, src, dst, null)
        }

        private fun debug() {
            val prevSecond = Date().seconds
            if (seconds != prevSecond) {
                Log.i(GIF_TAG, "draw:  cycle = " + cycle + " seconds = " + this.seconds)
                this.seconds = prevSecond
                cycle = 0
            }
            cycle += 1
        }

        private fun getMovieFromURL(src: String): Movie? {
            return try {
                val url = java.net.URL(src)
                val connection = url
                        .openConnection() as HttpURLConnection
                connection.doInput = true
                connection.connect()
                val input = connection.inputStream
                Movie.decodeStream(input)
            } catch (e: IOException) {
                e.printStackTrace()
                null
            }

        }

        override fun onCreate(surfaceHolder: SurfaceHolder) {
            super.onCreate(surfaceHolder)
            this.holder = surfaceHolder
        }

        override fun onDestroy() {
            super.onDestroy()
            handler.removeCallbacks(drawGIF)
            stopScheduledTask()
            stopIOSocket()
        }

        private fun stopScheduledTask() {
            if (mScheduledFutureStartOff != null) mScheduledFutureStartOff?.cancel(true)
            if (mScheduledFutureStartOn != null) mScheduledFutureStartOn?.cancel(true)
            mScheduler?.shutdown()
        }

        private fun stopIOSocket() {
            mSocket?.disconnect()
            mSocket?.off(SOCKET_EVENT_NAME)
        }

        override fun onVisibilityChanged(visible: Boolean) {
            this.visible = visible
            if (visible) {
                handler.post(drawGIF)
            } else {
                handler.removeCallbacks(drawGIF)
            }
        }

    }

    companion object {
        val SOCKET_EVENT_NAME = "gif"
        val CAPACITY = 8//WILL BECOME A CUSTOM PROPERTY
        val GIF_TAG = "SOCKET_EVENT_NAME"
        val DIMENSION = Pair(1, 3)
    }
}




